torchqdynamics/solver_utils.py

```diff
@@ -45,3 +45,14 @@ def bexpect(operators: Tensor, state: Tensor) -> Tensor:
     # TODO: Once QTensor is implemented, check if state is a density matrix or ket.
     # For now, we assume it is a density matrix.
     return torch.einsum('bij,...ji->...b', operators, state)
+
+
+def lindbladian(rho: torch.Tensor, H: torch.Tensor, L: torch.Tensor) -> torch.Tensor:
+    sum_nojump = (L.adjoint() @ L).sum(dim=0)
+
+    # non-hermitian Hamiltonian
+    H_nh = H - 0.5j * sum_nojump
+
+    H_nh_rho = H_nh @ rho
```

Comment:
Is there any performance gain by putting all of these in a single line? Since this will be a performance-sensitive part of the code (called at every time step for adaptive solvers), maybe we could use `torch.einsum` as much as possible, and group as much as we can in single calls.

Reply:
You're 100% right. (never be sorry for reviewing twice, it's better! 😋)

And **_arg_**, even for simple matmul it feels like we are going to need `einsum` everywhere 😭 
```python
import torch
import torch.utils.benchmark as benchmark

n = 50
b_H = 5
b_rho = 4
H = torch.rand(b_H, n, n, dtype=torch.complex64)
rho = torch.rand(b_H, b_rho, n, n, dtype=torch.complex64)
Hb = H[:, None, ...]  # (b_H, 1, n, n)

stmt1 = "Hb @ rho"
stmt2 = "torch.einsum('aij,a...jk->a...ik', H, rho)"

t1 = benchmark.Timer(stmt=stmt1, globals={'Hb': Hb, 'rho': rho})  # 402.55 us
t2 = benchmark.Timer(stmt=stmt2, globals={'H': H, 'rho': rho})  # 238.43 us

print(t1.timeit(100))
print(t2.timeit(100))
```

So I think we loose every time we augment a tensor to have a specific batching (e.g. everytime we do smthg like `self.H = H[:, None, ...]` and then use `H` in a matrix multiplication).

I propose to merge this PR, and then I'll do a new PR to einsum everything correctly. And then we absolutely need to look at https://optimized-einsum.readthedocs.io/en/stable/.

Reply:
On another note, I tried the pytorch sparse COO the other day. It gives some crazy GPU speedup on large matrices (x50 for N=1000), but supports very little actual useful operations. So I'm really convinced that a proper DIA matrix implementation will give us a crazy speedup.